### PR TITLE
Invalidate cart items when important props are updated

### LINKED ISSUE #16779: Cart expiry when changing product configurations
Hello,

I have a situation with Product Bundles which I originally directed to the SomewhereWarm team but they suggested it goes beyond PB and could affect other product types, such as Variable Products or Grouped Products.

My client uses Bundles as their main product, however change the items in each Bundle every week. This can cause a situation where a customer who has added the Bundle to their cart but not completed checkout, then comes back later (after the Bundle contents are changed) receives an error that one of the Bundled Items is not available (presumably because it has been removed from the Bundle). They have to remove the Product Bundle from their Cart and re-add it to continue.

There seems to be 2 ways of dealing with this:
1. Have an option with PB to allow updating of Bundles that may be sitting in carts when the content of such a Bundle is changed.
2. Have the ability to control Cart expiry time.

I approached the PB team about number 1, and have searched for solutions for number 2 myself, and have tried the code outlined in [this SO question](https://stackoverflow.com/questions/41167646/woocommerce-set-cart-expiration-interval) to do with wc_session_expiring and wc_session_expiration but ran into 2 issues:

- The mini-cart used in some themes (Storefront in my case) does not update to reflect the state of the Cart (empty) once it expires.  The mini-cart retains reference to the items that used to be in teh cart, until something new is added.
- In my testing this works for 'guests' (i.e. not-logged in users) but it does not work for logged-in users.

I am not an expert by any stretch of the imagination, and I apologise if I am bulldozing straight across something obvious - I have spent a long time trying to find an answer to this prior to coming here. Any thoughts or ideas will be gratefully accepted.

regards,
Sam

## Prerequisites

- [x ] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x ] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x ] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1.
2.
3.

## Expected/actual behavior

When I follow those steps, I see...

I was expecting to see...

## Isolating the problem

- [ ] This bug happens with only WooCommerce plugin active
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [ ] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
```
Copy and paste the system status report from **WooCommerce > System Status** in WordPress admin here.
```
</details>

### PR DESCRIPTION
**Pull request by mikejolley**
Adapted https://github.com/woocommerce/woocommerce/pull/18772 and fixes #16779

Adds a new function called `wc_get_cart_item_data_hash` which defines product properties that invalidate cart items. By default this includes product type and variation attributes.

If one of these props change, the hash will mismatch and a notice will show informing the customer of the change.

To test, add an item to the cart then edit it and change it’s product type. Visiting the cart page should show a new notice and the item will not be in your cart.

### CODE REVIEW

File: includes/wc-cart-functions.php [hunk 1]
```diff
@@ -481,3 +481,18 @@ function wc_shipping_methods_have_changed( $key, $package ) {
...
+ * Gets a hash of important product data that when changed should cause cart items to be invalidated.
+ *
+ * The woocommerce_cart_item_data_to_validate filter can be used to add custom properties.
+ *
+ * @param WC_Product $product Product object.
+ * @return string
+ */
+function wc_get_cart_item_data_hash( $product ) {
+	return md5( wp_json_encode( apply_filters( 'woocommerce_cart_item_data_to_validate', array(
+		'price'      => $product->get_type(),
```
Comment (sofyansitorus) [inline includes/wc-cart-functions.php]: @mikejolley The array key 'price' should be 'type' or 'product_type' instead.

**Review by sofyansitorus — CHANGES_REQUESTED** [review]

File: includes/wc-cart-functions.php (hunk 1 above)
Comment (kloon) [inline includes/wc-cart-functions.php]: I assume you meant to add a price and a type to the array here?

**Review by kloon — CHANGES_REQUESTED** [review]
Just that one issue in wc_get_cart_item_data_hash

File: includes/wc-cart-functions.php (hunk 1 above)
Comment (mikejolley) [reply includes/wc-cart-functions.php]: I don't think price should remove it from cart.

**Pull request merged**